seed/seedwriter/seed20.go

```diff
@@ -0,0 +1,316 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seedwriter
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/channel"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+var validSystemLabel = regexp.MustCompile("^[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*$")
+
+type policy20 struct {
+	model *asserts.Model
+	opts  *Options
+
+	warningf func(format string, a ...interface{})
+}
+
+var errNotAllowedExceptForDangerous = errors.New("cannot override channels, add local snaps or extra snaps with a model of grade higher than dangerous")
+
+func (pol *policy20) checkAllowedDangerous() error {
+	if pol.model.Grade() != asserts.ModelDangerous {
+		return errNotAllowedExceptForDangerous
+	}
+	return nil
+}
+
+func (pol *policy20) allowsDangerousFeatures() error {
+	return pol.checkAllowedDangerous()
+}
+
+func (pol *policy20) checkDefaultChannel(channel.Channel) error {
+	// TODO: consider allowing some channel overrides for >=signed
+	// Core 20 models?
+	return pol.checkAllowedDangerous()
+}
+
+func (pol *policy20) checkSnapChannel(ch channel.Channel, whichSnap string) error {
+	// TODO: consider allowing some channel overrides for >=signed
+	// Core 20 models?
+	return pol.checkAllowedDangerous()
+}
+
+func (pol *policy20) systemSnap() *asserts.ModelSnap {
+	return makeSystemSnap("snapd")
+}
+
+func (pol *policy20) modelSnapDefaultChannel() string {
+	// We will use latest/stable as default, model that want something else
+	// will need to to speficy a default-channel
+	return "latest/stable"
+}
+
+func (pol *policy20) extraSnapDefaultChannel() string {
+	// We will use latest/stable as default
+	// TODO: consider using just "stable" for these?
+	return "latest/stable"
+}
+
+func (pol *policy20) checkBase(info *snap.Info, availableSnaps *naming.SnapSet) error {
+	base := info.Base
+	if base == "" {
+		if info.GetType() != snap.TypeGadget && info.GetType() != snap.TypeApp {
```

Comment:
Should we use a whitelist here instead of a blacklist? Its a bit theoretical given that we hardly ever add new types but I wonder if `if info.GetType() == snap.TypeKernel || info.GetType() == snap.TypeBase || info.GetType() == snap.TypeOS` is more obvious ? Or maybe we could even grow something like `if info.GetType().NeedsBase()` or something?

Reply:
We have the conditions this way in:  https://github.com/snapcore/snapd/blob/master/overlord/snapstate/check_snap.go#L518

we have the other style (kind of) here, but there is also about default-providers:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L233

if we do something, it's probably for a follow up and we need to be consistent

Reply:
Right, :+1: for followup material then (not really that important too).